File: session-data-mongo/src/main/java/com/example/demo/DataInitializer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.example.demo;

import java.util.Arrays;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.event.ContextRefreshedEvent;
import org.springframework.context.event.EventListener;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;

/**
 *
 * @author hantsy
 */
@Component
@Slf4j
class DataInitializer {

    private final PostRepository posts;
    private final UserRepository users;
    private final PasswordEncoder passwordEncoder;

    public DataInitializer(PostRepository posts, UserRepository users, PasswordEncoder passwordEncoder) {
        this.posts = posts;
        this.users = users;
        this.passwordEncoder = passwordEncoder;
    }

    @EventListener(value = ContextRefreshedEvent.class)
    public void init() {
        initPosts();
        initUsers();
    }

    private void initUsers() {
        log.info("start users initialization  ...");
        this.users
            .deleteAll()
            .thenMany(
                Flux
                    .just("user", "admin")
                    .flatMap(
                        username -> {
                            List<String> roles = "user".equals(username)
                            ? Arrays.asList("ROLE_USER")
                            : Arrays.asList("ROLE_USER", "ROLE_ADMIN");

                            User user = User.builder()
                                    .roles(roles)
                                    .username(username)
                                    .password(passwordEncoder.encode("password"))
                                    .build();
                            return this.users.save(user);
                        }
                    )
            )
            .log()
            .subscribe(
                null,
                null,
                () -> log.info("done users initialization...")
            );
    }

    private void initPosts() {
        log.info("start data initialization  ...");
        this.posts
            .deleteAll()
            .thenMany(
                Flux
                    .just("Post one", "Post two")
                    .flatMap(
                        title -> this.posts.save(Post.builder().title(title).content("content of " + title).build())
                    )
            )
            .log()
            .subscribe(
                null,
                null,
                () -> log.info("done initialization...")
            );
    }

}

File: kotlin-gradle/src/main/kotlin/com/example/demo/DataInitializer.kt
package com.example.demo

import org.slf4j.LoggerFactory
import reactor.core.publisher.Flux
import javax.annotation.PostConstruct


class DataInitializr(val posts: PostRepository, val users: UserRepository) {
    private val log = LoggerFactory.getLogger(DataInitializr::class.java);

    // @EventListener(ContextRefreshedEvent::class)
    // @PostConstruct
    fun initData() {
        log.info("start data initialization ...")
        this.posts
                .deleteAll()
                .thenMany(
                        Flux
                                .just("Post one", "Post two")
                                .flatMap { it -> this.posts.save(Post(title = it, content = "content of " + it)) }
                )
                .log()
                .subscribe(
                        null,
                        null,
                        { log.info("done post initialization...") }
                )

        this.users
                .deleteAll()
                .thenMany(
                        Flux
                                .just("user", "admin")
                                .flatMap { it ->
                                    if (it == "user") this.users.save(User(username = it, password = "password", roles = listOf("ROLE_USER")))
                                    else this.users.save(User(username = it, password = "password", roles = listOf("ROLE_USER", "ROLE_ADMIN")))
                                }
                )
                .log()
                .subscribe(
                        null,
                        null,
                        { log.info("done users initialization...") }
                )
    }
}